### PR TITLE
Adding content on refactoring code for debugging and optimization / readability

### PR DESCRIPTION
**Pull request by avahoffman**
Adds to the course:

1. Introduction to Refactoring Code
   - Also includes a conceptual image
2. How AI can help with code refactoring
3. Example One: Correcting Syntax (indentations in python)
4. Example Two: Syntax Style and Convention
5. Example Three: Dry
6. Example Four: Keeping code alive

More in an additional PR!

### CODE REVIEW

**Comment by ehumph** [comment]
I **LOVE** how you arranged the chapter. The short examples are so easy to follow! I am going to use your approach to clean up the summarizing code chapter (and create the code writing chapter as well).

**Comment by avahoffman** [comment]
@ehumph glad you like it! I still have quite a bit to add, but I could definitely use a review soon :)

**Comment by ehumph** [comment]
@avahoffman if you think you've reached a good point, I can review tomorrow morning! I don't want to interrupt your flow (:

**Comment by avahoffman** [comment]
@ehumph sounds good! I'll start a new PR for other stuff.

File: 03-refactoring.Rmd [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    sum_result = calculate_sum(a, b)
+    print("Sum:", sum_result)
+    product_result = calculate_product(a, b)
+    print("Product:", product_result)
+
+
+if __name__ == '__main__':
+    main()
+```
+
```
Comment (ehumph) [inline 03-refactoring.Rmd:84]: Did ChatGPT just give you the corrected code, or did it also provide a written answer? I wonder if people will be taken aback by the fact there isn't a way (in our formatting) to tell the difference between the code you feed AI versus the code AI gives to you.

File: 03-refactoring.Rmd [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    print("Sum:", sum_result)
+
+if __name__ == '__main__':
+    main()
+```
+
+In this version, the `calculate_sum` function is simplified to take default arguments for a and b. This means the `default_values` function is no longer necessary. The a and b variables are also removed from `main`, as they are no longer needed. Much better and easier to look at overall.
+
+## Code optimization
+
```
Comment (ehumph) [inline 03-refactoring.Rmd:253]: I like loops! People will pry them from my cold dead hands (:

**Review by ehumph — APPROVED** [review]

File: 03-refactoring.Rmd (hunk 2 above)
Comment (avahoffman) [reply 03-refactoring.Rmd:253]: hahahahaa :P

File: 03-refactoring.Rmd [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+print_greetings()
+```
+
+Notice how repetitive this code is? `print()` and the string `"Hello, "` appear five times. We can improve this!
+
+:::{.query}
+"Could you please refactor my Python code to make it more DRY (Don't Repeat Yourself)?"
+:::
+
```
[the remark this answers is not in the document]
Comment (avahoffman) [reply 03-refactoring.Rmd:141]: yes, it understands. Will remove!

File: 03-refactoring.Rmd (hunk 1 above)
Comment (avahoffman) [reply 03-refactoring.Rmd:84]: This is a good point. Let me think about this.

File: 03-refactoring.Rmd [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+Code refactoring helps to reduce [technical debt](https://en.wikipedia.org/wiki/Technical_debt), which is the accumulation of development work that needs to be done in the future as a result of taking shortcuts or using less than optimal solutions. Refactoring can help to identify potential issues and bugs before they cause problems, resulting in more stable software. Finally, refactoring code to be more readable and less verbose can create fewer headaches for you in the future.
+
+To understand why code refactoring is important, imagine trying to cook in a disorganized and cluttered kitchen. You might be a fantastic cook, but you are spending more time searching for utensils and ingredients, and the quality of the meal could suffer as a result. Similarly, code that is disorganized and cluttered can be difficult to work with and can result in lower-quality software. Refactoring is like cleaning and organizing the kitchen - it makes the process of cooking (or coding) smoother and more efficient.
+
```
[the remark this answers is not in the document]
Comment (avahoffman) [reply 03-refactoring.Rmd:17]: ChatGPT came up with it :)

File: 03-refactoring.Rmd (hunk 1 above)
Comment (ehumph) [reply 03-refactoring.Rmd:87]: Ooh, yes, that works really well!

**Comment by ehumph** [comment]
@avahoffman Go ahead and merge! I am restructuring the code summarization chapter to follow your formatting and style, but having the old stuff up for a bit is no problem at all.

File: 03-refactoring.Rmd (hunk 1 above)
Comment (avahoffman) [reply 03-refactoring.Rmd:87]: Cool! You can prompt it with 
```
:::{.ai_response}
Text and code here
:::

**Pull request merged**